pep-0704.rst

```diff
@@ -0,0 +1,137 @@
+
+Further, activation of virtual environment uses a slightly different syntax/mechanism to activate on different platforms.
+
+It also creates a scope for mistakes, since users must remember to activate the virtual environment before running an installer like ``pip``. On certain Linux distributions, this can result in the installer modifying files that are owned by the operating system (this is partially mitigated by :pep:`668` for distributions that opt-in to marking their environments accordingly).
+
+This PEP proposes that the default behaviour of installers like ``pip`` should be to require a virtual environment to be active. This would make it easier for new users to get started with Python, and would reduce the scope for mistakes.
+
+
+Specification
+=============
+
+When a user runs an installer without an active virtual environment, the installer MUST print an error message and exit with a non-zero exit code. The error message MUST inform the user that a virtual environment is required, SHOULD provide shell-specific instructions on how to create and activate a virtual environment named ``.venv``, and SHOULD provide a link to a documentation page that explains how to create and activate a virtual environment.
+
+The installer SHOULD also provide a mechanism to disable this behaviour, allowing a user to opt-in to running the tool outside of a virtual environment. The installer MAY include this in the error message that it prints, and MUST include it in the documentation page it links to.
+
+If the installer includes a link to a documentation page in the error message described above, the documentation page SHOULD be written with a focus on new users. It should explain what virtual environments are, why they are required, and how to create and activate a virtual environment. It SHOULD include instructions for the most common shells and platforms.
+
+Installers MAY choose to implement this default behaviour on any Python versions, but MUST implement it on Python 3.13 or newer.
+
+
+Backwards Compatibility
+=======================
+
+This PEP is backwards incompatible with workflows where users are using installers outside of virtual environments. Such users will be prompted with an error message and will need to either:
+
+- explicitly opt-in to running the installer outside of a virtual environment, or
+- create and use a virtual environment
+
+Users who are already using virtual environments will not be affected by this change.
+
+Workflow tools (which manage virtual environments for the user, under the hood) should be unaffected, since they should already be using the virtual environment for running the installer.
+
+
+Security Implications
+=====================
+
+This PEP does not introduce any new security implications.
```

Comment:
You could potentially argue that it could improve security slightly, as any malicious package will (at least initially) only be installed into a single isolated virtualenv, as opposed to the system environment used for all projects (perhaps even with administrator rights).

Reply:
There's roughly no security win, since the ability to execute arbitrary code means you more or less have all the powers the user does.

Reply:
> There's roughly no security win, since the ability to execute arbitrary code means you more or less have all the powers the user does.

Yeah, but it _does_ avoid the "admin escape hatch" that several users have learned to take if they cannot install something with their regular user, and it turns out they're trying to install to the system python. It's great if this is not done by default anymore, but it still helps even if you're accidentally installing something as root user to not mess up your system python, but put them in a separate env (though when you're installing something as root, obviously any potential security improvements aren't relevant anymore).

Reply:
> Yeah, but it does avoid the "admin escape hatch" that several users have learned to take if they cannot install something with their regular user, and it turns out they're trying to install to the system python. 

We've closed that with EXTERNALLY-MANAGED (PEP 668).